### DISCUSSION TITLE
Rendering (point clouds, particles) as points

### DISCUSSION THREAD

**Discussion by totalgee** [StereoKit/StereoKit#436] [Q&A] [answered]
Hi, StereoKit looks really nice, congratulations. I was looking at the API and haven't found a way to render large numbers of points (assuming one provides a suitable custom shader). Is this possible, e.g. to draw a mesh in "points mode" so only point primitives are emitted (rather than triangles)? Or alternatively to provide a buffer of vertices and directly render them as points (as in Vulkan's `VK_PRIMITIVE_TOPOLOGY_POINT_LIST` or OpenGL's `GL_POINTS` draw primitive)?

It would be useful for drawing large numbers of (possibly transparent) particles or point cloud datasets, although perhaps this doesn't fit as well into a "solid objects" PBR system. Or maybe it's already possible and I've not yet come across it in the docs...

**Comment by maluoi** [comment, marked as answer]
Point rendering isn't built in! But it is relatively straightforward with a bit of shader work.

There's a chunk of demo code with a small helper class that does point cloud rendering over here:
https://github.com/StereoKit/StereoKit/blob/master/Examples/StereoKitTest/Demos/DemoPointCloud.cs

That code relies on this shader:
https://github.com/StereoKit/StereoKit/blob/master/Examples/Assets/point_cloud.hlsl

And you can see it all in action over here!
https://twitter.com/koujaku/status/1506066374073847811

This code basically creates a quad for each point, and then expands that quad in the shader, using perspective or screen-space depending on the settings. It's a pretty fast and generally reliable way to draw points!

**Reply by totalgee** [reply]
Thanks for the pointer to that example!
